pkg/migrate/cstor/volume.go

```diff
@@ -0,0 +1,902 @@
+			Desired:     version.Current(),
+			AutoUpgrade: false,
+			Status: cstor.VersionStatus{
+				Current:            version.Current(),
+				DependentsUpgraded: true,
+			},
+		}
+		_, err = v.OpenebsClientset.CstorV1().CStorVolumeConfigs(v.OpenebsNamespace).
+			Create(cvcObj)
+		if err != nil {
+			return err
+		}
+	}
+	klog.Info("Patching the target svc with cvc owner ref")
+	err = v.patchTargetSVCOwnerRef()
+	if err != nil {
+		errors.Wrap(err, "failed to patch cvc owner ref to target svc")
+	}
+	return nil
+}
+
+func (v *VolumeMigrator) patchTargetSVCOwnerRef() error {
+	svcObj, err := v.KubeClientset.CoreV1().
+		Services(v.OpenebsNamespace).
+		Get(v.PVName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cvcObj, err := v.OpenebsClientset.CstorV1().
+		CStorVolumeConfigs(v.OpenebsNamespace).
+		Get(v.PVName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	newSVCObj := svcObj.DeepCopy()
+	newSVCObj.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(cvcObj,
+			cstor.SchemeGroupVersion.WithKind(cvcKind)),
+	}
+	data, err := GetPatchData(svcObj, newSVCObj)
+	if err != nil {
+		return err
+	}
+	_, err = v.KubeClientset.CoreV1().
+		Services(v.OpenebsNamespace).
+		Patch(v.PVName, types.StrategicMergePatchType, data)
+	return err
+}
+
+// updateStorageClass recreates a new storageclass with the csi provisioner
+// the older annotations with the casconfig are also preserved for information
+// as the information about the storageclass cannot be gathered from other
+// resources a temporary storageclass is created before deleting the original
+func (v *VolumeMigrator) updateStorageClass(pvName, scName string) error {
+	klog.Infof("Updating storageclass %s with csi parameters", scName)
+	scObj, err := v.KubeClientset.StorageV1().
+		StorageClasses().
+		Get(scName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
+	if scObj == nil || scObj.Provisioner != cstorCSIDriver {
+		tmpSCObj, err := v.createTmpSC(scName)
+		if err != nil {
+			return err
+		}
+		replicaCount, err := v.getReplicaCount(pvName)
+		if err != nil {
+			return err
+		}
+		cspcName, err := v.getCSPCName(pvName)
+		if err != nil {
+			return err
+		}
+		csiSC := tmpSCObj.DeepCopy()
+		csiSC.ObjectMeta = metav1.ObjectMeta{
+			Name:        scName,
+			Annotations: tmpSCObj.Annotations,
+		}
+		csiSC.Provisioner = cstorCSIDriver
+		csiSC.AllowVolumeExpansion = &trueBool
+		csiSC.Parameters = map[string]string{
+			"cas-type":         "cstor",
+			"replicaCount":     replicaCount,
+			"cstorPoolCluster": cspcName,
+		}
+		if scObj != nil {
+			err = v.KubeClientset.StorageV1().
+				StorageClasses().Delete(scName, &metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		scObj, err = v.KubeClientset.StorageV1().
+			StorageClasses().Create(csiSC)
+		if err != nil {
+			return err
+		}
+		v.StorageClass = scObj
+		err = v.KubeClientset.StorageV1().
```

Comment:
If pod restarts over here then there  `temporary storage class` will not be deleted. So good to move this deletion out of  block i.e outof `if scObj == nil || scObj.Provisioner != cstorCSIDriver {}`